core/drivers/crypto/caam/ae/caam_ae_gcm.c

```diff
@@ -0,0 +1,354 @@
+	/*
+	 * 128 bits Ctri value
+	 */
+	uint32_t yi[4];
+	/*
+	 * 128 bits Ctr0 value
+	 */
+	uint32_t y0[4];
+};
+
+/*
+ * Update of the cipher operation of complete block except
+ * last block. Last block can be partial block.
+ *
+ * @caam_ctx AE Cipher context
+ * @encrypt  Encrypt or decrypt direction
+ * @src      Source data to encrypt/decrypt
+ * @dst      [out] Destination data encrypted/decrypted
+ * @final    Last block flag
+ */
+static bool caam_ae_do_block_gcm(struct caam_ae_ctx *caam_ctx, bool encrypt,
+				 struct caamdmaobj *src, struct caamdmaobj *dst,
+				 bool final)
+{
+	/*
+	 * CAAM Errata:
+	 * When running GCM when the IV is not 12 bytes (96 bits),
+	 * it is possible to "roll over" the 32-bit counter value.
+	 * When this occurs (unless counter starts at -1),
+	 * the AES operation will generate an incorrect MAC.
+	 * This occurs even when -1 is used as the counter for the last block.
+	 * The problem is that the 32-bit counter will overflow into the h
+	 *  value, thus corrupting the MAC.
+	 * In order to reliably work around this issue,
+	 * the AES operation must be stopped after initialization to
+	 * determine the counter value to see whether/when it will roll over.
+	 * Then, before the offending block is processed,
+	 * the context needs to be saved. The one block gets processed twice :
+	 * GCM, restore MAC, GMAC over its ciphertext,
+	 * then patch up the message and AAD lengths, and carry on as normal.
+	 */
+	uint32_t *desc = NULL;
+	struct gcm_caam_ctx_layout ctx = {};
+	size_t input_length = 0;
+	uint32_t processed_blocks = 0;
+	uint32_t counter_value = 0;
+	uint32_t corrupted_block_size = 0;
+	uint32_t yi_1s_complement = 0;
+	uint32_t remaining_len = 0;
+
+	if (!caam_ctx)
+		return false;
+
+	desc = caam_ctx->descriptor;
+
+	/*
+	 *  for encrypt:
+	 *  1) Run GCM until we get to the block which will
+	 *     roll over the counter.
+	 *  2) Save the current ghash value
+	 *  3) Encrypt that one block (creating a bad hash value)
+	 *  4) Restore the hash value
+	 *  5) Save current AAD len
+	 *  6) Run ciphertext of the block in as AAD
+	 *  7) Restore the AAD len
+	 *  8) Run GCM on the rest of the message
+	 *  9) Compute and store the MAC/tag
+	 *
+	 *  for decrypt:
+	 *  1) Run GCM until we get to the block which will
+	 *     roll over the counter.
+	 *  2) Save the current ghash value
+	 *  3) Decrypt that one block (creating a bad hash value)
+	 *  4) Restore the hash value
+	 *  5) Save current AAD len
+	 *  6) Run ciphertext of the block in as AAD
+	 *  7) Restore the AAD len
+	 *  8) Run GCM on the rest of the message
+	 *  9) Compute and store the MAC/tag
+	 */
+
+	if (!src || src->orig.length == 0 ||
+	    caam_ctx->nonce.length == AES_GCM_DEFAULT_NONCE_LEN)
+		return false;
+
+	memcpy(&ctx, caam_ctx->ctx.data, sizeof(struct gcm_caam_ctx_layout));
+	processed_blocks = TEE_U32_BSWAP(ctx.yi[3]);
+	input_length = src->orig.length;
+	counter_value = processed_blocks + DIV_ROUND_UP(input_length, 16);
```

Comment:
`ROUNDUP()`?

Reply:
Please check this https://github.com/OP-TEE/optee_os/pull/7116#discussion_r1840996456

Reply:
@etienne-lms, why is `DIV_ROUND_UP()` preferred over `ROUNDUP()` here?

Reply:
It seems to the counter computation needs the number of 16 bytes blocks in `input_length`.
[`((input_length + 15) >> 4)`](https://github.com/OP-TEE/optee_os/pull/7116#discussion_r1840996456) is what `DIV_ROUND_UP(input_length, 16)` computes.

Reply:
Thanks, I missed that part. Next question, why `DIV_ROUND_UP()` and not `ROUNDUP_DIV()`? It seems the latter is more safe.

Reply:
Right, `ROUNDUP_DIV()` would be better.

Reply:
I will do the change. Thanks.